Apiserver dedups owner references before creating/updating/patching objects

roycaihw opened the pull request:
<!--  Thanks for sending a pull request!  Here are some tips for you:

1. If this is your first time, please read our contributor guidelines: https://git.k8s.io/community/contributors/guide/first-contribution.md#your-first-contribution and developer guide https://git.k8s.io/community/contributors/devel/development.md#development-guide
2. Please label this pull request according to what type of issue you are addressing, especially if this is a release targeted pull request. For reference on required PR/issue labels, read here:
https://git.k8s.io/community/contributors/devel/sig-release/release.md#issuepr-kind-label
3. Ensure you have added or ran the appropriate tests for your PR: https://git.k8s.io/community/contributors/devel/sig-testing/testing.md
4. If you want *faster* PR reviews, read how: https://git.k8s.io/community/contributors/guide/pull-requests.md#best-practices-for-faster-reviews
5. If the PR is unfinished, see how to mark it: https://git.k8s.io/community/contributors/guide/pull-requests.md#marking-unfinished-pull-requests
-->

**What type of PR is this?**
/kind bug

**What this PR does / why we need it**:
apiserver allows objects being created/updated with duplicate owner reference entries, which may lead to huge objects being created in the system.

As discussed in the previous SIG meeting (Nov 4th). The API server adds a warning header and dedups  without failing requests. Duplication is determined over the entire entry, not just the UID field.

**Which issue(s) this PR fixes**:
<!--
*Automatically closes linked issue when PR is merged.
Usage: `Fixes #<issue number>`, or `Fixes (paste link of issue)`.
_If PR is about `failing-tests or flakes`, please post the related issues/tests in a comment and do not use `Fixes`_*
-->
Fixes #96065

**Special notes for your reviewer**:
~This doesn't resolve https://github.com/kubernetes/kubernetes/issues/96065#issuecomment-721339514. Today the [field manager](https://github.com/kubernetes/kubernetes/blob/ac996a37f65a8f11640808a883e9522832ed364f/staging/src/k8s.io/apiserver/pkg/endpoints/handlers/create.go#L164) ignores any [preprocessing](staging/src/k8s.io/apiserver/pkg/registry/rest/create.go) the apiserver does before creating/updating an object. cc @apelisse @jennybuckley~ As discussed in the SIG meeting, the deduping happens before we write managed fields. 

**Does this PR introduce a user-facing change?**:
<!--
If no, just write "NONE" in the release-note block below.
If yes, a release note is required:
Enter your extended release note in the block below. If the PR requires additional action from users switching to the new release, include the string "action required".

For more information on release notes see: https://git.k8s.io/community/contributors/guide/release-notes.md
-->
```release-note
Duplicate owner reference entries in create/update/patch requests now get deduplicated by the API server. The client sending the request now receives a warning header in the API response. Clients should stop sending requests with duplicate owner references. The API server may reject such requests as early as 1.24. 
```

/sig api-machinery
/assign @lavalamp @deads2k

fedebongio commented:
/triage accepted

roycaihw commented:
/hold 

How to fix this issue may be controversial. If we did it right we should validate and reject requests with duplicate owner references. Let's discuss the options in https://github.com/kubernetes/kubernetes/issues/96065

dims commented:
/uncc

roycaihw commented on staging/src/k8s.io/apiserver/pkg/endpoints/handlers/rest.go:
```diff
@@ -53,6 +55,8 @@ const (
 	// Everyone chooses 30.
 	requestTimeout = 34 * time.Second
+	// TODO(roycaihw): stop serving requests with duplicate owner reference entries in 1.24
+	DuplicateOwnerReferencesWarning = ".metadata.ownerReferences contains duplicate entries; API server dedups owner references in 1.20+, rejects such requests in 1.24+; please fix your requests"
```
the dedup behavior is deprecated. Rejecting requests in 1.24 gives us one year to remove this behavior

https://kubernetes.io/docs/reference/using-api/deprecation-policy/#deprecating-a-feature-or-behavior

roycaihw commented:
/hold cancel

I updated this PR based on our discussion in the previous SIG meeting. Please take a look

cc @lavalamp @deads2k

roycaihw commented on staging/src/k8s.io/apiserver/pkg/endpoints/handlers/create.go:
```diff
@@ -152,6 +153,15 @@ func createHandler(r rest.NamedCreater, scope *RequestScope, admit admission.Int
 			)
 		}
+		if accessor, accessorErr := meta.Accessor(obj); accessorErr == nil {
+			// Dedup owner references before updating managed fields
```
I did not put this before creating admission attributes because I don't think any admission plugin cares about the duplication, so let's keep it as it is. 

I don't think we should put this before audit logging because audit log can help identify problematic requests.

deads2k commented on staging/src/k8s.io/apiserver/pkg/endpoints/handlers/create.go:
```diff
@@ -152,6 +153,15 @@ func createHandler(r rest.NamedCreater, scope *RequestScope, admit admission.Int
+			refs := accessor.GetOwnerReferences()
+			deduped, lastDuplicate := dedupOwnerReferences(refs)
+			if lastDuplicate != nil {
+				warning.AddWarning(req.Context(), "", fmt.Sprintf(DuplicateOwnerReferencesWarningFormat, lastDuplicate.UID))
```
it seems like it would be nicer to list as many of the duplicates as you could.

roycaihw replied:
The reason was we have some [limits](https://github.com/kubernetes/kubernetes/blob/1df1f882c4421d313abb468fef5c15f89983471c/staging/src/k8s.io/apiserver/pkg/endpoints/filters/warning.go#L39-L42) that could truncate the warning message. I will change to list as many of the duplicates as we can

roycaihw replied:
done

deads2k commented on staging/src/k8s.io/apiserver/pkg/endpoints/handlers/create.go:
```diff
@@ -152,6 +153,15 @@ func createHandler(r rest.NamedCreater, scope *RequestScope, admit admission.Int
 				options,
 			)
 		}
+		if accessor, accessorErr := meta.Accessor(obj); accessorErr == nil {
```
Please add comment indicating that this must come after admission so that we properly de-dupe ownerreferences added by mutating admission.

roycaihw replied:
This line happens before mutating admission. I added another dedup call after mutating admission

The orders are:
- CREATE: dedup -> field manager -> admission -> dedup
- UPDATE: dedup -> (field manager -> admission -> dedup) * N
- PATCH: (apply patch -> field manager -> admission -> dedup) * N

For patch requests a bad request can still wipes out managed fields. And it's not easy to pass the request context down to each patcher.

deads2k commented on staging/src/k8s.io/apiserver/pkg/endpoints/handlers/rest.go:
```diff
@@ -53,6 +55,12 @@ const (
+	// warning message from getting truncated. We do provide the UID of the last duplicate entry we seen as a hint, to
+	// help the client identify and fix the request.
+	// TODO(roycaihw): stop serving requests with duplicate owner reference entries in 1.24
+	DuplicateOwnerReferencesWarningFormat = ".metadata.ownerReferences contains duplicate entries, last duplicate UID: %v; API server dedups owner references in 1.20+, rejects such requests in 1.24+; please fix your requests"
```
rather than promise us to a future release, perhaps, "may reject such requests as early as 1.24"?

roycaihw replied:
done

deads2k commented:
This is a bug fix and it's as easy on clients as we can make it.  Please add a test to ensure it works for CRs (I see one for configmaps now) and I'm happy with it modulo my comments.

/milestone v1.20

roycaihw commented on staging/src/k8s.io/apiserver/pkg/endpoints/handlers/patch.go:
```diff
@@ -569,9 +569,14 @@ func (p *patcher) patchResource(ctx context.Context, scope *RequestScope) (runti
 
 	wasCreated := false
-	p.updatedObjectInfo = rest.DefaultUpdatedObjectInfo(nil, p.applyPatch, p.applyAdmission)
+	p.updatedObjectInfo = rest.DefaultUpdatedObjectInfo(nil, p.applyPatch, p.applyAdmission, dedupOwnerReferencesTransformer)
```
managed fields get updated in `applyPatch`. Currently a bad request can still wipe out managed fields.

apelisse replied:
interesting how you mention that at the exact same time that this happened: https://github.com/kubernetes/kubernetes/issues/96688

apelisse replied:
cc @kwiesmueller

kwiesmueller replied:
Wouldn't it be enough to also dedup before apply?
Well... we'd have to dedup on the patchObj we decode into unstructured I think.

In general, I think Apply would still be broken of sorts until we make all the fields you equal on keys.
This could even solve both problems because at least for apply the request would be rejected on duplicate keys?
Correct me if I'm wrong @apelisse

apelisse replied:
Apply does the schema validation so there's no way you're going to be able to apply duplicated keys. If you apply a key that is already in the object, it's going to be merged with the applied key. And, unfortunately, if the object has duplicated keys, I don't know if it can be applied.

roycaihw replied:
here I'm thinking of: 
- if a patch appends a duplicate owner ref, the managedFields update will fail

mutating admission is not involved. But I agree #96688 is also a problem

kwiesmueller replied:
Yeah what I meant is, that we should make all fields in the ownerRef keys in the schema so we replicate the same behavior?

roycaihw replied:
> Apply does the schema validation so there's no way you're going to be able to apply duplicated keys

That's good news. So the problem I was thinking of https://github.com/kubernetes/kubernetes/pull/96185#discussion_r526397606 only happens to the other patchers (json patch, json merge patch, SMP)

roycaihw commented:
> add a test to ensure it works for CRs

added

roycaihw committed:
apiserver dedups owner references and adds warning

for CREATE and UPDATE requests, we check duplication before managedFields
update, and after mutating admission; for PATCH requests, we check
duplication after mutating admission

deads2k commented:
/lgtm
/approve
/hold

holding for a squash to 3-ish commits and for discussion on the managed fields issue.  I don't think that this makes the existing problem substantively worse, but if it does we'll need to wait to merge.

roycaihw committed:
unit and integration tests

apiserver dedups and adds warning in CREATE/UPDATE/PATCH requests;
also handles duplication caused by mutating admission.

roycaihw committed:
generated

jeremyrickard commented:
/test pull-kubernetes-e2e-kind

roycaihw committed:
make sure managedFields are written

roycaihw commented on test/integration/apiserver/apiserver_test.go:
```diff
@@ -2216,19 +2216,26 @@ func TestDedupOwnerReferences(t *testing.T) {
+			// wipe out managed fields. Note that this happens to built-in resources but
+			// not custom resources. In future we should either dedup before writing manage
+			// fields, or stop deduping and reject the request.
+			// assertManagedFields(t, dependent)
```
Two behaviors regarding managedFields that surprised me, but may be expected. FYI @apelisse 
1. once an object has empty managedFields, the field manager won't write managedFields for future requests made by this client, even if those requests don't contain duplicate owner refs. 
2. without this PR, the field manager still succeeded to write `managedFields` for custom resources, even when duplicate owner ref existed in the request. Note that `"f:ownerReferences"` in the custom resource doesn't contain UIDs. 

with this PR, for built-in resource:
```
# create with dup owner refs
managedFields: [{apiserver.test Update v1 2020-11-18 19:44:36 -0800 PST FieldsV1 {"f:metadata":{"f:ownerReferences":{".":{},"k:{\"uid\":\"03d3ba47-a815-46aa-84c7-70ee450b7761\"}":{".":{},"f:apiVersion":{},"f:kind":{},"f:name":{},"f:uid":{}}}}}}]
# update with dup owner refs
managedFields: [{apiserver.test Update v1 2020-11-18 19:44:36 -0800 PST FieldsV1 {"f:metadata":{"f:ownerReferences":{".":{},"k:{\"uid\":\"03d3ba47-a815-46aa-84c7-70ee450b7761\"}":{".":{},"f:apiVersion":{},"f:kind":{},"f:name":{},"f:uid":{}}}}}}]
# patch, append a dup owner ref
fieldmanager.go:186] [SHOULD NOT HAPPEN] failed to update managedFields for /, Kind=: failed to convert new object (/v1, Kind=ConfigMap) to smd typed: .metadata.ownerReferences: duplicate entries for key [uid="03d3ba47-a815-46aa-84c7-70ee450b7761"]
managedFields: []
# update, append a different owner ref
managedFields: []
# patch, append a different owner ref
managedFields: []
```
for custom resource (note the UID isn't written in the managed fields): 
```
# repeat all the steps above, the managed fields always read:
[{apiserver.test Update cr.bar.com/v1 2020-11-18 19:44:36 -0800 PST FieldsV1 {"f:metadata":{"f:ownerReferences":{}}}}]
```
without this PR, the behavior of built-in resources will be:
```
# create with dup owner refs
fieldmanager.go:186] [SHOULD NOT HAPPEN] failed to update managedFields for /, Kind=: failed to convert new object (/v1, Kind=ConfigMap) to smd typed: .metadata.ownerReferences: duplicate entries for key [uid="29b54e45-e089-4b12-bfd6-6ab63f20204a"]
managedFields: []
# repeat all the remaining steps:
managedFields: []
```
and the behavior of custom resources won't change.

kwiesmueller replied:
I didn't expect the managedFields to stay empty. Does this also happen on apply?

apelisse replied:
> 1. once an object has empty managedFields, the field manager won't write managedFields for future requests made by this client, even if those requests don't contain duplicate owner refs. 

That's because we don't track fields for objects that don't have managed fields, until they are server-side applied. That's to prevent existing objects from suddenly becoming bigger (because of the managed-fields) on existing clusters.
When we wipe, we could decide that we need to keep tracking, but I'd say this is WAI.

> 2. without this PR, the field manager still succeeded to write `managedFields` for custom resources, even when duplicate owner ref existed in the request. Note that `"f:ownerReferences"` in the custom resource doesn't contain UIDs. 

it looks like ownerReferences is considered atomic for custom-resources. Can you look at the CRD and say if the ownerReference list is associative?

roycaihw replied:
> but I'd say this is WAI.

Thanks for the explanation! That makes sense. 

> it looks like ownerReferences is considered atomic for custom-resources

IIUC the field manager uses the CRD schema to determine that. The [CRD used](https://github.com/kubernetes/kubernetes/blob/18099e1ef7283d9ab09c45c5a4a90e26fdce1161/test/integration/etcd/data.go#L593-L606) doesn't have schema. I also verified that API server doesn't insert the in-tree object meta schema into CRD. I think this is expected.

apelisse replied:
> The CRD used doesn't have schema.

We treat all lists as atomic for schema-less CRDs, all good! 👍

roycaihw reviewed (commented):
@apelisse As we spoke offline, I updated the test to make sure managedFields is non-empty. PTAL.

The managedFields-related behavior is improved with this PR. We can file issue to track the remaining pre-existing problems. The main purpose of this PR is to prevent object with duplicate owner refs written to etcd.

roycaihw commented:
/test pull-kubernetes-conformance-kind-ga-only-parallel

deads2k commented:
/lgtm

apelisse commented on test/integration/apiserver/apiserver_test.go:
```diff
@@ -2136,3 +2140,208 @@ func expectPartialObjectMetaV1EventsProtobuf(t *testing.T, r io.Reader, values .
+
+			klog.Infof("patching dependent with duplicate owner reference")
+			dependent = c.patchDependentWithOwner(dependent, fakeRefA)
+			// TODO: currently a patch request that duplicates owner references can still
```
I think that comment is correct and makes sense! 👍

roycaihw commented:
/hold cancel

The remaining comments are not blockers. Thanks for reviewing

roycaihw commented:
/retest
/priority critical-urgent
